Ability to specify temp dir to gather_analytics()

tchellomello opened the pull request:
##### SUMMARY
For large environments configured with a small /tmp partition,
gather_analytics() tends to fail due to space constraints.
This patch allows the ability to configure a temporary staging
directory to be used.

Thanks to Gabe Muniz @gamuniz (:1st_place_medal:) for the collaboration on this patch.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
awx: 11.2.0
```

##### ADDITIONAL INFORMATION


<!--- Paste verbatim command output below, e.g. before and after your change -->
```
[root@awx ~]# ls -lad /gather2
ls: cannot access /gather2: No such file or directory

[root@awx ~]# grep ^ANALYT /etc/tower/conf.d/insights.py 
ANALYTICS_GATHER_DIR = '/gather2/testing/this/does/not/exist'

[root@awx ~]# awx-manage  gather_analytics
Last analytics run was: 2020-05-14 04:21:14.542282+00:00
/gather2/testing/this/does/not/exist/0885970a-3653-4d39-9d1e-501701c1ee41_2020-05-14-160833+0000.tar.gz

[root@awx ~]# tree /gather2
/gather2
└── testing
    └── this
        └── does
            └── not
                └── exist
                    └── 0885970a-3653-4d39-9d1e-501701c1ee41_2020-05-14-160833+0000.tar.gz
5 directories, 1 file
```

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -97,6 +97,9 @@ def gather(dest=None, module=None, collection_type='scheduled'):
         from awx.main.analytics import collectors
         module = collectors
 
+    tgz_tempdir = getattr(settings, 'ANALYTICS_GATHER_DIR', None)
```
Whenever there's a new setting introduced like this, we generally also put a matching default definition somewhere in `awx/settings/defaults.py` e.g.,

```python
settings.ANALYTICS_GATHER_DIR = None
```

ryanpetrello replied:
Also, cc @gamuniz you're making a new directory here, but it's not actually being used anywhere below that I can tell?

Maybe below we should do something like:

```python
dest = dest or tgz_tempdir or tempfile.mkdtemp(prefix='awx_analytics')
```

tchellomello replied:
@ryanpetrello thanks for your comments. PR updated

ryanpetrello reviewed (changes requested):
This needs some additional work.

wenottingham commented:
Thinking about this, we may want a specific "scratch space" setting that encompasses both this and the per-job-run project checkouts - cc @ryanpetrello @AlanCoding

gamuniz commented:
@ryanpetrello I agree with notting here maybe we should make this a more broad setting since we are removing the project/inventory blockers which could cause even more disk space issues, unless I am misreading his comment.

ryanpetrello commented:
I mean, for jobs, at least, what you're referring to is `TMPDIR` :)

tchellomello committed:
Ability to specify temp dir to gather_analytics()

For large environments configured with a small /tmp partition,
gather_analytics() tends to fail due to space constraints.
This patch allows the ability to configure a temporary staging
directory to be used.

Thanks to Gabe Muniz for the collaboration on this patch.

Signed-off-by: Marcelo Moreira de Mello <tchello.mello@gmail.com>

wenottingham commented:
> I mean, for jobs, at least, what you're referring to is `TMPDIR` :)

But for analytics, it's the same thing. So the right answer is to probably have a setting that exposes TMPDIR to all tasks like this.

ryanpetrello reviewed (changes requested):
I agree with @wenottingham that the _correct_ way to accomplish this is to stop solving it one-off, and introduce a new mechanism for controlling `TMPDIR` in *all* background task execution.

ryanpetrello commented:
Spoke with @tchellomello and we're going to discuss what https://github.com/ansible/awx/pull/7025#pullrequestreview-413824605 means and what a PR might look like.  He's going to file an issue that describes the problem.
